src/sql/magic_cmd.py

```diff
@@ -51,6 +52,14 @@ def _validate_execute_inputs(self, line):
             "explore",
             "snippets",
         ]
+        COMMANDS_CONNECTION_REQUIRED = [
+            "tables",
+            "columns",
+            "test",
+            "profile",
+            "explore",
+        ]
+        COMMANDS_SQLALCHEMY_ONLY = ["tables", "columns", "test", "explore"]
```

Comment:
are we sure that `explore` doesn't work with a dbapi connection? can you try removing the error and see if it works? I quickly looked at the code and I didn't see anything that makes it sqlalchemy-only

Reply:
This is the whole traceback I get when I comment `util.support_only_sql_alchemy_connection("%sqlcmd")` to not raise the error and run 

```
import duckdb
%load_ext sql
conn = duckdb.connect()
%sql conn
%sqlcmd explore --table penguins.csv
```

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In[7], line 1
----> 1 get_ipython().run_line_magic('sqlcmd', 'explore --table penguins.csv')

File [~/miniconda3/envs/jupysql/lib/python3.10/site-packages/IPython/core/interactiveshell.py:2417](https://file+.vscode-resource.vscode-cdn.net/Users/sanggyuan/Desktop/intern/jupysql/~/miniconda3/envs/jupysql/lib/python3.10/site-packages/IPython/core/interactiveshell.py:2417), in InteractiveShell.run_line_magic(self, magic_name, line, _stack_depth)
   2415     kwargs['local_ns'] = self.get_local_scope(stack_depth)
   2416 with self.builtin_trap:
-> 2417     result = fn(*args, **kwargs)
   2419 # The code below prevents the output from being displayed
   2420 # when using magics with decodator @output_can_be_silenced
   2421 # when the last Python token in the expression is a ';'.
   2422 if getattr(fn, magic.MAGIC_OUTPUT_CAN_BE_SILENCED, False):

File [~/Desktop/intern/jupysql/src/sql/magic_cmd.py:89](https://file+.vscode-resource.vscode-cdn.net/Users/sanggyuan/Desktop/intern/jupysql/~/Desktop/intern/jupysql/src/sql/magic_cmd.py:89), in SqlCmdMagic._validate_execute_inputs(self, line)
     80         raise exceptions.RuntimeError(
     81             f"Cannot execute %sqlcmd {command} because there "
     82             "is no active connection. Connect to a database "
     83             "and try again."
     84         )
     86     # if command in COMMANDS_SQLALCHEMY_ONLY:
     87     #     util.support_only_sql_alchemy_connection("%sqlcmd")
---> 89     return self.execute(command, others)
     90 else:
     91     raise exceptions.UsageError(
     92         f"%sqlcmd has no command: {command!r}. "
     93         "Valid commands are: {}".format(
     94             ", ".join(AVAILABLE_SQLCMD_COMMANDS)
     95         )
     96     )

File [~/Desktop/intern/jupysql/src/sql/magic_cmd.py:116](https://file+.vscode-resource.vscode-cdn.net/Users/sanggyuan/Desktop/intern/jupysql/~/Desktop/intern/jupysql/src/sql/magic_cmd.py:116), in SqlCmdMagic.execute(self, cmd_name, others, cell, local_ns)
    114 cmd = router.get(cmd_name)
    115 if cmd:
--> 116     return cmd(others)

File [~/Desktop/intern/jupysql/src/sql/cmd/explore.py:23](https://file+.vscode-resource.vscode-cdn.net/Users/sanggyuan/Desktop/intern/jupysql/~/Desktop/intern/jupysql/src/sql/cmd/explore.py:23), in explore(others)
     20 parser.add_argument("-t", "--table", type=str, help="Table name", required=True)
     21 args = parser.parse_args(others)
---> 23 table_widget = TableWidget(args.table)
     24 display(table_widget)

File [~/miniconda3/envs/jupysql/lib/python3.10/site-packages/ploomber_core/telemetry/telemetry.py:644](https://file+.vscode-resource.vscode-cdn.net/Users/sanggyuan/Desktop/intern/jupysql/~/miniconda3/envs/jupysql/lib/python3.10/site-packages/ploomber_core/telemetry/telemetry.py:644), in Telemetry.log_call.._log_call..wrapper(*args, **kwargs)
    642             result = func(_payload, *args, **kwargs)
    643     else:
--> 644         result = func(*args, **kwargs)
    645 except Exception as e:
    646     metadata_error = {
    647         # can we log None to posthog?
    648         "type": getattr(e, "type_", None),
   (...)
    651         **_payload,
    652     }

File [~/Desktop/intern/jupysql/src/sql/widgets/table_widget/table_widget.py:41](https://file+.vscode-resource.vscode-cdn.net/Users/sanggyuan/Desktop/intern/jupysql/~/Desktop/intern/jupysql/src/sql/widgets/table_widget/table_widget.py:41), in TableWidget.__init__(self, table)
     38 html_style = utils.load_css(f"{BASE_DIR}[/css/tableWidget.css](https://file+.vscode-resource.vscode-cdn.net/css/tableWidget.css)")
     39 self.add_to_html(html_style)
---> 41 self.create_table(table)
     43 # register listener for jupyter lab
     44 self.register_comm()

File [~/Desktop/intern/jupysql/src/sql/widgets/table_widget/table_widget.py:60](https://file+.vscode-resource.vscode-cdn.net/Users/sanggyuan/Desktop/intern/jupysql/~/Desktop/intern/jupysql/src/sql/widgets/table_widget/table_widget.py:60), in TableWidget.create_table(self, table)
     56 """
     57 Creates an HTML table with default data
     58 """
     59 rows_per_page = 10
---> 60 rows, columns = fetch_sql_with_pagination(table, 0, rows_per_page)
     61 rows = parse_sql_results_to_json(rows, columns)
     63 query = f"SELECT count(*) FROM {table}"

File [~/Desktop/intern/jupysql/src/sql/util.py:302](https://file+.vscode-resource.vscode-cdn.net/Users/sanggyuan/Desktop/intern/jupysql/~/Desktop/intern/jupysql/src/sql/util.py:302), in fetch_sql_with_pagination(table, offset, n_rows, sort_column, sort_order)
    294 query = f"""
    295 SELECT * FROM {table} {order_by}
    296 OFFSET {offset} ROWS FETCH NEXT {n_rows} ROWS ONLY"""
    298 rows = ConnectionManager.current.execute(query).fetchall()
    300 columns = ConnectionManager.current.raw_execute(
    301     f"SELECT * FROM {table} WHERE 1=0"
--> 302 ).keys()
    304 return rows, columns

AttributeError: 'duckdb.DuckDBPyConnection' object has no attribute 'keys'
```